### DISCUSSION TITLE
--innodb-optimize-keys=AFTER_IMPORT_PER_TABLE → errno: 150 "Foreign key constraint is incorrectly formed"

### DISCUSSION THREAD

**Discussion by grinapo** [mydumper/mydumper#1211] [General]
Have ben hunting this problem for a few days now, and neither `mariadb` nor `myloader` was particularly helpful. The schema import fails, which cause the data import fail, for a specific table, referenced by a bunch of others:

```
2023-07-14 05:32:57 [ERROR] - Thread 3: Error restoring: Can't create table `dbname`.`fieldname` (errno: 150 "Foreign key constraint is incorrectly formed")
```

The generic advice is to check the types (they match), check whether it is either primary key or indexed first key (it is), and that's basically all. Even `SHOW ENGINE INNODB STATUS` (last foreign key error) wasn't particularly telling.

All in all, turned out that this is caused by `--innodb-optimize-keys=AFTER_IMPORT_PER_TABLE` arg during myloading. Funny thing is, after checking the documentation that this is _also supposed to be the default_.

Removing it seems to be 100% removing the error. Kind of logical, if it would work the way that `myloader` first creates the table without indexen, then import data, then create indexes, and mariadb (this one's 10.4.25) clearly dislike creating a table referenced by a FK with a visibly non-indexed field.

I fixed mine so I'm okay, but for the rest… 
- why does it break, and how not to?
- why breaks as an arg if that's the default?
- maybe the doc (myloader --help) shall be extended a bit about warning lusers about unexpected consequences?

**Comment by jdevinemt** [comment]
I just ran into this issue myself. The only thing I have to add is that this error did not occur several months ago when running the same command. Unfortunately, I don't know the exact version I used where the error did not occur, but I do remember it was right around when v0.19 of mydumper was released.

**Reply by jdevinemt** [reply]
Although this error didn't occur in the past, it being triggered is probably the correct behavior. I just used `SHOW ENGINE INNODB STATUS` on the target database to inspect the error further. In my case it had to do with the referenced column not having an index where it is the first key.

I may have upgraded my database up one minor version of MariaDB since last time I ran this migration, so that might have been why the error didn't occur previously.

In my case, I think myloader is functioning as intended.

**Comment by davidducos** [comment]
Hi @grinapo / @jdevinemt , sorry that I miss this thread 2 years ago.
When we use `--innodb-optimize-keys/--optimize-keys`, myloader should be splitting the table if possible, into 3 pieces:
- Create table with only the columns and the primary key
- Secondary indexes, this could be multiple statements. 
- Foreign key

Then the logic will be:
- Create the table
- import the data
- create secondary indexes
- create foreign key

Take into account that we could create multiple tables at the same time and the data is import as soon as the table is created, secondary indexes starts depending on the option in `--innodb-optimize-keys/--optimize-keys` but the data import for the table that we are going to create the index has been completed, and finally the foreign keys are created after all the indexes has been created (https://github.com/mydumper/mydumper/blob/master/src/myloader/myloader.c#L553). 

BTW, I reviewed the logic of myloader of the latest version.

**Reply by jdevinemt** [reply]
Thanks David. As always, I appreciate your dedication to this project.

I had a bunch of mismatching foreign keys where the the referenced column didn't have the exact same data type as the foreign key column. I fixed all those foreign key columns to match their referenced column data type. However, I still received the error when attempting to migrate the database again (I use the stream option to pipe the output of mydumper into myloader). 

I believe the reason I still got these foreign key warnings was because I had not dropped the schemas in the target database prior to running the migration. Since the default --optimize-keys value is AFTER_IMPORT_PER_TABLE, I think it was referencing the existing referencing table, which probably hadn't been dropped and recreated yet.

I think it would be better practice on my part to make sure the target database is empty before I run a complete database migration like this. But it brings up two questions.

1) How come AFTER_IMPORT_PER_TABLE is the default value for the --optimize-keys setting instead of AFTER_IMPORT_ALL_TABLES? I assume it's because this option raises key related errors earlier on rather than waiting until the end of data population, therefore providing faster feedback?
2) Would adding an --overwrite-schemas option be a good idea or even a possibility? It's not a big deal for me to handle this myself, but it would be helpful. However, there may not be that many people that use multiple schemas in their database. The only reason I have multiple schemas is because I am working with a legacy system.

Neither of these questions are pressing. But I would be curious to hear your opinion.

**Reply by davidducos** [reply]
Hi @jdevinemt, myloader should be restoring the backup with `foreign_key_checks = OFF`, so you should not see any referential error. If you do, then, you are configuring mydumper incorrectly. How? well, if you are using the --defaults-file you might be missing some options that you can find in /etc/mydumper.cnf like: 
```
[myloader_session_variables]
# These are the defaults
SQL_MODE='NO_AUTO_VALUE_ON_ZERO' /*!40101
UNIQUE_CHECKS=0 /*!40114
FOREIGN_KEY_CHECKS=0 /*!40114
# SQL_LOG_BIN will be added by default in next release and -e will be deprecated
SQL_LOG_BIN=0
```
depending on your db vendor. As you can see there, the `FOREIGN_KEY_CHECKS=0`. 
I think that we need to confirm this before continue with other analysis.

**Reply by jdevinemt** [reply]
Hey @davidducos, I am using the unmodified default configuration file and it matches the configuration you provided. Is it possible the `foreign_key_checks` option doesn't control whether the `errno: 150 "Foreign key constraint is incorrectly formed"` is displayed? I always assumed the `foreign_key_checks` option only controlled whether it enforced the existence of the value in the referenced column.

**Reply by davidducos** [reply]
Hi @jdevinemt, I might misunderstood your last comment. I thought that you have also issues with the data, but seems that you only have issues with the table structures. 
Anyways, take into account that FOREIGN_KEY_CHECKS also affects the DDL that  you can send, so you shouldn't see any issue. 
I will be reviewing the code, as will be useful to see the command that is failing

**Reply by jdevinemt** [reply]
Hey @davidducos, here is the command I am running.

```sh
mydumper -h $SOURCE_DB_HOST -P $SOURCE_DB_PORT -u $SOURCE_DB_USER -p $SOURCE_DB_PASSWORD --triggers --routines --skip-definer \
    --complete-insert --database $DATABASES --compress-protocol --stream --threads 8 \
  | myloader -h $TARGET_DB_HOST -P $TARGET_DB_PORT -u $TARGET_DB_USER -p $TARGET_DB_PASSWORD --overwrite-tables \
    --compress-protocol --stream --threads 8
```

The issue only occurs when the schemas already exist in the target database. If I drop the schemas in the target database before running this command, there is no issue. I also tried the command above with `--optimize-keys AFTER_IMPORT_ALL_TABLES`, but the issue still occurs.

Here is an example of the error `** (myloader:1145): CRITICAL **: 12:19:15.075: Thread 11 using connection 59 - ERROR 1005: Can't create table xxx.xxx (errno: 150 "Foreign key constraint is incorrectly formed")`

**Reply by davidducos** [reply]
Can you share the table structure or a test case so I can troubleshoot this bug?

**Reply by jdevinemt** [reply]
Here is the sanitized table structure.

```mysql
# taken from myloader output
CREATE TABLE `TableA` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `rev_id` int(10) unsigned NOT NULL,
  `modified` datetime NOT NULL,
  `name` varchar(550) COLLATE utf8_unicode_ci,
  PRIMARY KEY (`id`,`rev_id`) USING BTREE
) ENGINE=InnoDB AUTO_INCREMENT=37 DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci

# taken from DDL (in IDE)
# I removed some unrelated columns for the sake of brevity
create table TableB
(
    id           int unsigned auto_increment primary key,
    a_rev_id     int unsigned not null,
    date         date         null,
    constraint TableB_ibfk_2 foreign key (a_rev_id) references TableA (rev_id)
)
create index a_id
    on TableB (a_rev_id);
```

Here is the output of `SHOW ENGINE INNODB STATUS;`:

```
------------------------
LATEST FOREIGN KEY ERROR
------------------------
2025-05-15 18:19:15 0x400048ec4910 Error in foreign key constraint of table schema/TableB:
there is no index in referenced table which would contain
the columns as the first columns, or the data types in the
referenced table do not match the ones in table. Constraint:
,
  CONSTRAINT `TableB_ibfk_2` FOREIGN KEY (`a_rev_id`) REFERENCES `TableA` (`rev_id`)
The index in the foreign key in table is a_id
Please refer to https://mariadb.com/kb/en/library/foreign-keys/ for correct foreign key definition.
```

My only hunch is that is has to do with the fact that before running the migration, TableB and its indexes already exists (from previous run of the migration), which is causing the error to trigger when one of the myloader threads tries to drop and create TableA before TableB is dropped.

If it helps at all, my target database is a MariaDB 10.5 AWS RDS database. The version of mydumper is `mydumper v0.19.1-3, built against MySQL 8.0.41-32 with SSL support`.  The only other warning in the log is `Set session failed: SET SESSION SQL_LOG_BIN = 0` (probably because it's an RDS database), but I do already have the sync_binlog set to 0.

**Reply by davidducos** [reply]
Hi @jdevinemt,
Mmmmm... if there isn't an index on TableA over rev_id, which is requirement for MariaDB to create the constraint, this is not a myloader bug or issue, this is something that you should fix on your database to perform the migration. BTW, is your db vendor and version the same on source and target? if the answer is "no", then you need to fix the incompatibilities first, and if it is yes, then it will be strange to see how this was working before without the index on TableA.

**Reply by jdevinemt** [reply]
Hey @davidducos, the following indexes exists on TableA. I didn't include them in my last comment, because they don't exist at the time myloader is creating the the table since it defaults to adding the indexes after the table is created.

```mysql
create index name on TableA (name(255));

create index rev_id on TableA (rev_id);
```

I am migrating from MariaDB 10.1 to 10.5, but the rev_id index exists in the source database.

Do you think the error is occuring because TableB, its foreign key and indexes already exists (from a previous migration run) when myloader tries to recreate TableA on the subsequent migration?

Again, I can easily get around this by dropping the schema these tables belong to before re-running the migration.

**Reply by davidducos** [reply]
Ah, now makes sense, if dropping the whole database before running myloader makes it works, then yeah!
Please, let me know if you need further assistance.

**Reply by jdevinemt** [reply]
Is there any chance you might add a `--drop-schemas` option to myloader? It's not a big deal to do this manually, but it could be a handy option.

I think the only reason the foreign key error is being raised in this case is because the TableA uses a composite key and TableB references the second column in that composite key for its foreign key column. This issue shouldn't occur if tables and their associations are designed well, unlike the tables I'm working with :P

**Reply by davidducos** [reply]
Hi @jdevinemt , I think that this is an opportunity to change -o/--overwrite-tables to -o/--drop-tables, and make it consistent with --drop-schemas. We can also allow --drop-tables to accept the parameters in --purge-mode and remove it.

**Reply by davidducos** [reply]
I think that we can close this as we already have https://github.com/mydumper/mydumper/issues/1878

**Reply by jdevinemt** [reply]
Definitely. Thanks!
